architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,329 @@
+
+----
+There are also a few major decisions to be made at a top-down level, which are captured here.
+
+#### User-defined types are declarative
+Our scenario doc and jobs-to-be-done analysis highlights that authoring user-defined-types needs to be easy and have a low barrier to entry. This design prop supports that goal by providing a fully-declarative path for implementing a user-defined type.
+
+- Validation is declarative (OpenAPI with no imperative component).
+- Versioning and conversion is declarative (CEL with no imperative component).
+- Asynchronous processing uses Recipes.
+
+The alternative would be to implement a resource provider, *which is already supported*. However, we're not going to push users towards implementing a resource provider.
+
+##### Advantages (of each option considered)
+
+The advantage of a fully-declarative system is that easy to implement and that there are no additional control-plane components to host. Users can easily reuse the Recipe engine, which expected to be common. If we consider the need to support a variety of cloud technologies as dependencies, it's clear that Recipes are a key scenario.
+
+--- 
+The advantage of the resource provider model is that if gives users full-control over the resource lifecycle (both request-path and async processing).
+
+##### Disadvantages (of each option considered)
+
+The disadvantages of a fully-declarative system is that it's not as powerful. Users can only do the things we enable. 
+
+As a comparison we can consider the Kubernetes extensibility model, because it's mature and widely used. The Kubernetes extensibility model requires users to implement a control-plane component (webhook) when versioning is needed, or when imperative validation is needed. The downside of this is that a misbehaving webhook can impact the availability of the overall system. Kubernetes is adding [declarative validation](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/) using CEL so that imperative validation is needed in fewer scenarios. 
+
+---
+The disadvantage of the resource provider model is that it's very complex and requires a lot of code. It takes an engineer familiar with Radius approximately a week of time to implement a new type. Additionally, the resource provider is a control-plane component that users must manage. 
+
+There's a significant cost to us to because we need to create SDKs and documentation that make users successful at writing a resource provider.
+
+##### Proposed Option
+
+The proposed option is that we lean heavily on declarative functionality. We'll support two models, but push users towards the declarative model:
+- Fully declarative (OpenAPI, CEL, Recipes)
+- Fully imperative (Resource Provider)
+
+If we need to, we can add more hooks in the future to create hybrids. This is complex and costly so we should set a high bar for it.
+
+#### User-defined types are implemented as an RP 
+
```

Comment:
Feedback from review: make it clear that this is about the implementation, not what users implement.